batch/batch/worker/worker.py

```diff
@@ -171,6 +174,68 @@ def compose(auth: Union[MutableMapping, str, bytes], registry_addr: str = None):
 image_lock = aiorwlock.RWLock()
 
 
+async def read_file(fs: LocalAsyncFS, path: str) -> str:
+    if os.path.exists(path):
+        return (await fs.read(path)).decode()
+    return ''
+
+
+class ResourceUsageMonitor:
+    VERSION = 1
+
+    def __init__(self, container_name: str, output_file_path: str, create_time_secs: int):
+        self.container_name = container_name
+        self.output_file_path = output_file_path
+
+        self.last_time_ns: Optional[int] = int(create_time_secs * 10**3)
+        self.last_cpu_ns: Optional[int] = 0  # assume zero CPU usage at container start time
+
+        self.task: Optional[asyncio.Future] = None
+
+    def write_header(self):
+        with open(self.output_file_path, 'w') as f:
+            f.write(f'#version={self.VERSION}' + '\n')
+            f.write('\t'.join(['time_msecs', 'cpu_usage', 'memory_usage_bytes']) + '\n')
+
+    def cpu_ns(self) -> int:
+        with open(f'/sys/fs/cgroup/cpu/{self.container_name}/cpuacct.usage', 'r') as f:
+            return int(f.read().rstrip())
+
+    def percent_cpu_usage(self) -> float:
+        now_time_ns = time_ns()
+        now_cpu_ns = self.cpu_ns()
+
+        assert self.last_time_ns is not None and self.last_cpu_ns is not None
+
+        cpu_usage = (now_cpu_ns - self.last_cpu_ns) / (now_time_ns - self.last_time_ns)
+
+        self.last_time_ns = now_time_ns
+        self.last_cpu_ns = now_cpu_ns
+
+        return cpu_usage
+
+    def memory_usage_bytes(self) -> Optional[int]:
+        with open(f'/sys/fs/cgroup/memory/{self.container_name}/memory.usage_in_bytes', 'r') as f:
+            return int(f.read().rstrip())
+
+    async def measure(self):
+        data = [time_msecs(), self.percent_cpu_usage(), self.memory_usage_bytes()]
+        data = [str(d) for d in data]
+        with open(self.output_file_path, 'a') as f:
```

Comment:
After this merges, we should peek at the workers and make sure their CPU use under heavy load is not too bad. Opening and closing files takes some time. If we have 64 jobs in parallel that's ~12 files per second. I feel like that should be OK, but it would be good to check that a fully loaded worker still monitors at a reasonable pace.

Reply:
Would it be better to have an open file handle?

Reply:
Actually, the reason why I didn't have an open file handle is we need it to flush to disk after every line is written. I can directly add a flush.